File: app/src/main/java/de/kuschku/libquassel/syncables/types/invokers/IBacklogManager.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.syncables.types.invokers;

import android.support.annotation.NonNull;

import java.util.List;

import de.kuschku.libquassel.exceptions.SyncInvocationException;
import de.kuschku.libquassel.functions.types.SyncFunction;
import de.kuschku.libquassel.message.Message;
import de.kuschku.libquassel.syncables.types.interfaces.QBacklogManager;

public class IBacklogManager implements Invoker<QBacklogManager> {
    @NonNull
    private static final IBacklogManager invoker = new IBacklogManager();

    private IBacklogManager() {
    }

    @NonNull
    public static IBacklogManager get() {
        return invoker;
    }

    @Override
    public void invoke(SyncFunction function, QBacklogManager obj) throws SyncInvocationException {
        switch (function.methodName) {
            case "receiveBacklog": {
                obj._receiveBacklog((int) function.params.get(0), (int) function.params.get(1), (int) function.params.get(2), (int) function.params.get(3), (int) function.params.get(4), (List<Message>) function.params.get(5));
            }
            break;
            case "receiveBacklogAll": {
                obj._receiveBacklogAll((int) function.params.get(0), (int) function.params.get(1), (int) function.params.get(2), (int) function.params.get(3), (List<Message>) function.params.get(4));
            }
            break;
            case "update": {
                InvokerHelper.update(obj, function.params.get(0));
            }
            break;
            default: {
                throw new SyncInvocationException(function.className + "::" + function.methodName);
            }
        }
    }
}

File: lib/src/main/java/de/kuschku/libquassel/quassel/syncables/interfaces/IBacklogManager.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2020 Janne Mareike Koschinski
 * Copyright (c) 2020 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.quassel.syncables.interfaces

import de.justjanne.libquassel.annotations.ProtocolSide
import de.justjanne.libquassel.annotations.SyncedCall
import de.justjanne.libquassel.annotations.SyncedObject
import de.kuschku.libquassel.protocol.BufferId
import de.kuschku.libquassel.protocol.MsgId
import de.kuschku.libquassel.protocol.QuasselType
import de.kuschku.libquassel.protocol.QVariantList
import de.kuschku.libquassel.protocol.QtType
import de.kuschku.libquassel.protocol.qVariant

@SyncedObject(name = "BacklogManager")
interface IBacklogManager : ISyncableObject {
  @SyncedCall(target = ProtocolSide.CORE)
  fun requestBacklog(
    bufferId: BufferId,
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0
  ) {
    sync(
      target = ProtocolSide.CORE,
      "requestBacklog",
      qVariant(bufferId, QuasselType.BufferId),
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
    )
  }

  @SyncedCall(target = ProtocolSide.CORE)
  fun requestBacklogFiltered(
    bufferId: BufferId,
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    type: Int = -1,
    flags: Int = -1
  ) {
    sync(
      target = ProtocolSide.CORE,
      "requestBacklogFiltered",
      qVariant(bufferId, QuasselType.BufferId),
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(type, QtType.Int),
      qVariant(flags, QtType.Int),
    )
  }

  @SyncedCall(target = ProtocolSide.CORE)
  fun requestBacklogAll(
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0
  ) {
    sync(
      target = ProtocolSide.CORE,
      "requestBacklogAll",
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
    )
  }

  @SyncedCall(target = ProtocolSide.CORE)
  fun requestBacklogAllFiltered(
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    type: Int = -1,
    flags: Int = -1
  ) {
    sync(
      target = ProtocolSide.CORE,
      "requestBacklogAll",
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(type, QtType.Int),
      qVariant(flags, QtType.Int),
    )
  }

  @SyncedCall(target = ProtocolSide.CLIENT)
  fun receiveBacklog(
    bufferId: BufferId,
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    messages: QVariantList
  ) {
    sync(
      target = ProtocolSide.CLIENT,
      "receiveBacklog",
      qVariant(bufferId, QuasselType.BufferId),
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(messages, QtType.QVariantList),
    )
  }

  @SyncedCall(target = ProtocolSide.CLIENT)
  fun receiveBacklogFiltered(
    bufferId: BufferId,
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    type: Int = -1,
    flags: Int = -1,
    messages: QVariantList
  ) {
    sync(
      target = ProtocolSide.CLIENT,
      "receiveBacklogFiltered",
      qVariant(bufferId, QuasselType.BufferId),
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(type, QtType.Int),
      qVariant(flags, QtType.Int),
      qVariant(messages, QtType.QVariantList),
    )
  }

  @SyncedCall(target = ProtocolSide.CLIENT)
  fun receiveBacklogAll(
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    messages: QVariantList
  ) {
    sync(
      target = ProtocolSide.CLIENT,
      "receiveBacklogAll",
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(messages, QtType.QVariantList),
    )
  }

  @SyncedCall(target = ProtocolSide.CLIENT)
  fun receiveBacklogAllFiltered(
    first: MsgId = MsgId(-1),
    last: MsgId = MsgId(-1),
    limit: Int = -1,
    additional: Int = 0,
    type: Int = -1,
    flags: Int = -1,
    messages: QVariantList
  ) {
    sync(
      target = ProtocolSide.CLIENT,
      "receiveBacklogAllFiltered",
      qVariant(first, QuasselType.MsgId),
      qVariant(last, QuasselType.MsgId),
      qVariant(limit, QtType.Int),
      qVariant(additional, QtType.Int),
      qVariant(type, QtType.Int),
      qVariant(flags, QtType.Int),
      qVariant(messages, QtType.QVariantList),
    )
  }
}
